### PR TITLE
Fix indexing inside SplineX2X when outputing orbitals fewer than it holds

### PR DESCRIPTION
**Pull request by camelto2**
Please review the [developer documentation](https://github.com/QMCPACK/qmcpack/wiki/Development-workflow)
on the wiki of this project that contains help and requirements.

## Proposed changes

Found a segfault when trying to use hybridrep in orbital optimization. The source of the problem is that when doing orbital optimization, the underlying orbital set includes unoccupied orbitals. However, if we are calling the evaluations from inside a DiracDeterminant, the ValueVector psi size is ways set to the number of occupied orbitals, not the full size of the SPOSet. The fix here allows for proper indexing into psi, and this is actually producing a successful hyrbid rep optimization. 

The spline classes do not properly guard against accessing out of bounds for the ValueVector psi, and I expect this is more general problem for all SPOSets. 

TO DO:
- [x] Fix SplineR2R
- [x] Fix SplineC2C
- [x] add useful unit tests
- [x] Fix SplineC2R 
- [x] Fix SplineC2COMPTarget
- [x] Fix SplineC2ROMPTarget

## What type(s) of changes does this code introduce?
_Delete the items that do not apply_

- Bugfix

### Does this introduce a breaking change?
- No

## What systems has this change been tested on?
M1 mac

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes. This PR is up to date with current the current state of 'develop'
- Yes. Code added or changed in the PR has been clang-formatted
- Yes. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- No. Documentation has been added (if appropriate)

### CODE REVIEW

**Comment by camelto2** [comment]
The current units tests that are failing, are actually failing as expected. This is due to other spline implementations needing the same fixes. I'll work on those next and get the test passing

**Comment by camelto2** [comment]
As is, the unit tests are modified to allow for the (requested number of orbitals) < (the number of orbitals defined in the SPOSet). The current unit test in test_hybridrep only checks SplineC2C and SplineR2R. These modifications to the unit test fail with an out of bounds access without the changes to SplineC2C and SplineR2R. The various assign_x functions in the splines now guard against out of bounds access for the requested number of orbitals, and then the tests pass

The diamond2x1x1 test can be modified in the same way to test SplineC2R, but I'm still taking a look at that. For now, I think we could just merge this, and do subsequent PRs on SplineC2R and SplineC2ROMPTarget/SplineC2COMPTarget if those need it

File: src/QMCWaveFunctions/BsplineFactory/SplineC2C.cpp
Comment (ye-luo) [inline src/QMCWaveFunctions/BsplineFactory/SplineC2C.cpp:171]: Should be called last_cplx as the last complex output orbital.

File: src/QMCWaveFunctions/BsplineFactory/SplineC2C.cpp
Comment (ye-luo) [inline src/QMCWaveFunctions/BsplineFactory/SplineC2C.cpp:424]: last_cplx here.

File: src/QMCWaveFunctions/BsplineFactory/SplineC2C.cpp
Comment (ye-luo) [inline src/QMCWaveFunctions/BsplineFactory/SplineC2C.cpp:347]: last_cplx here.

File: src/QMCWaveFunctions/BsplineFactory/SplineC2C.cpp
Comment (ye-luo) [inline src/QMCWaveFunctions/BsplineFactory/SplineC2C.cpp:260]: last_cplx here.

File: src/QMCWaveFunctions/BsplineFactory/SplineC2C.cpp
Comment (camelto2) [reply src/QMCWaveFunctions/BsplineFactory/SplineC2C.cpp]: done

File: src/QMCWaveFunctions/BsplineFactory/SplineC2C.cpp
Comment (camelto2) [reply src/QMCWaveFunctions/BsplineFactory/SplineC2C.cpp]: done

File: src/QMCWaveFunctions/BsplineFactory/SplineC2C.cpp
Comment (camelto2) [reply src/QMCWaveFunctions/BsplineFactory/SplineC2C.cpp]: done

File: src/QMCWaveFunctions/BsplineFactory/SplineC2C.cpp
Comment (camelto2) [reply src/QMCWaveFunctions/BsplineFactory/SplineC2C.cpp]: done

**Review by ye-luo — DISMISSED** [review]

**Comment by prckent** [comment]
Test this please

**Comment by camelto2** [comment]
Well the V100 Clang offload tests were failing, which makes me think the tests I modified actually end up going through the SplineC2COMPTarget, so i need to include those fixes in this PR as well. So I went ahead and fixed SplineC2R and modified a test for that as well since it seems all the Spline classes need a fix

**Comment by camelto2** [comment]
Ok, I think all tests should pass now

**Comment by ye-luo** [comment]
Will approve and merge once the CI update is merged.

**Comment by ye-luo** [comment]
Test this please

**Review by ye-luo — APPROVED** [review]

**Pull request merged**
